[documentation] How to debug memory leaks (#1904)
reactor/reactor-netty: docs/asciidoc/faq.adoc

Reactor Netty adds information for the connection at the beginning of every log record (when this is possible).
Use the id of the connection in order to extract all log records for a particular HTTP request.
For more information see <<faq.logging-prefix>>

[[faq.memory-leaks]]
== How can I debug a memory leak?

> By default, Reactor Netty uses a direct memory as this is more performant

Reviewer: Remove "a" from "a direct memory".

> when there are many native I/O operations (working with sockets) as this can remove the copying operations.

Reviewer: Add a comma after "(working with sockets)".

> As allocation and deallocation are expensive operations, the buffers (used in Reactor Netty by default) are pooled objects.

Reviewer 2: I would actually suggest to turn this sentence to:
```suggestion
As allocation and deallocation are expensive operations, Reactor Netty also uses pooled buffers by default.
```

> As allocation and deallocation are expensive operations, the buffers, used in Reactor Netty by default, are pooled objects.

Reviewer: Change the commas around "used in Reactor Netty by default" to parentheses. While technically correct, with other commas nearby, it is hard to parse. Replacing those two commas with parentheses will clarify the sentence.

> For more information see https://github.com/netty/netty/wiki/Reference-counted-objects[Reference Counted Objects].

Reviewer: Add a comma after "information".

> In order to be able to debug memory issues with the direct memory and the pooled buffers, Netty provides a special memory leak detection mechanism.

Reviewer: Remove "In order". This is a general rule. It is nearly always unnecessary.

> Follow the instructions https://github.com/netty/netty/wiki/Reference-counted-objects#troubleshooting-buffer-leaks[Troubleshooting Buffer Leaks]

Reviewer: Add "for" between "instructions" and the link.

> in order to enable this mechanism. In addition to the instructions provided by Netty, Reactor Netty provides a special

Reviewer: Remove "in order".

> logger `_reactor.netty.channel.LeakDetection` that helps to identify where the memory leak might be located inside Reactor Netty

Reviewer: Add parentheses around `_reactor.netty.channel.LeakDetection`.

or whether Reactor Netty already forwarded the ownership of the buffers to the application/framework.
By default, this logger is disabled. To enable it, increase the log level to `DEBUG`.

> 

Reviewer 3: In addition I think it has a sense to mention reducing memory: lesser memory application has sooner memory leak will occur.

Some time ago I used
`-Djdk.nio.maxCachedBufferSize` and `-XX:MaxDirectMemorySize` for it when debug gateway.

Also there is an article https://dzone.com/articles/troubleshooting-problems-with-native-off-heap-memo

Author: @reactor/netty-team @nkonev I added a note for reducing the used memory. Please check.

> By default, this logger is disabled, in order to enable it - increase the log level to `DEBUG`.

Reviewer: Change to

By default, this logger is disabled. To enable it, increase the log level to `DEBUG`.
